Reviews (everyone push to this branch)

asmeurer opened the pull request:
So from now to May 22 (Sunday), we should do the first round of reviews. The goal here is to make the paper coherent. 

**Everyone who wants to be an author needs to do this.**

What you should do now is read through the paper and make any changes you see fit, and push them to this branch (`reviews`). If you do not have push access, let me know and I can add you.  

Everyone pushing to the same branch will minimize merge conflicts. You should constantly be pulling in changes to avoid conflicts, and commit and push as soon as you make a change.

If you don't have time to do a full review, partial reviews are still helpful, and you can also look at the TODO.md file at the root of this repo to see some things that need to be fixed.

asmeurer committed:
Clear some trialing whitespace

asmeurer committed:
Remove redundant sentence

asmeurer committed:
Fix the ordering of sections in the introduction

asmeurer commented:
To clarify, this is both for small changes (grammar fixes, adding citations, etc.), and larger changes (rearranging paragraphs/sections, rewriting sections if necessary). Basically, anything that you see as necessary to make the paper high quality and coherent.

hargup committed:
Use a more neutral language

Signed-off-by: Harsh Gupta <gupta.harsh96@gmail.com>

* Use the term Free and Open Source instead of just Open Source.
* Remove value judgement about the permissiveness of the BSD license

asmeurer committed:
Clean up a strangely worded sentence.

asmeurer committed:
variables -> symbols

asmeurer committed:
Merge branch 'reviews' of github.com:sympy/sympy-paper into reviews

hargup commented on introduction.tex:
```diff
@@ -1,11 +1,11 @@
 % cite BSD?
 SymPy was started by Ond\v{r}ej \v{C}ert\'{\i}k in 2005, and it has since
-grown into a large open source project, with over 500 contributors. SymPy is
+grown into a large project with over 500 contributors. SymPy is
```
Mentioning open source here was redundant.

asmeurer committed:
Example is supposed to be for the whole paragraph, not just that sentence

hargup committed:
Minor spelling and grammer fixes

Signed-off-by: Harsh Gupta <gupta.harsh96@gmail.com>

asmeurer committed:
Lead in the example

asmeurer committed:
Merge branch 'reviews' of github.com:sympy/sympy-paper into reviews

asmeurer committed:
Avoid first person

asmeurer committed:
Rewrite a sentence for clarity

asmeurer committed:
Remove "just"

This detracts from the point. Two structurally equal expressions are
necessarily mathematically equal, but the point is that == isn't intended to
be mathematical at all.

asmeurer committed:
Functions are called, not their arguments

asmeurer committed:
Add a bit to a sentence

asmeurer committed:
Small grammar fixes

asmeurer committed:
The advantages aren't exactly "listed" above

asmeurer committed:
integer -> integral

asmeurer committed:
Use the same variable as the example

asmeurer committed:
A symbol named t

asmeurer commented:
Any suggestions on how to fix the non-wrapping line here (top of page 4)?

<img width="802" alt="screenshot 2016-05-17 16 30 28" src="https://cloud.githubusercontent.com/assets/71486/15338187/c2eb80b2-1c4c-11e6-8794-a29f7e347ac2.png">

Is it OK to hyphenate the texttt text? That doesn't even look [so easy to do](https://tex.stackexchange.com/questions/44361/how-to-automatically-hyphenate-within-texttt).

asmeurer committed:
Add some extra words, to fix the long line

asmeurer commented:
Nevermind, I changed it to "some of the" and it wraps now :)

certik commented:
Yeah, it probably had a few words on the line, most of them long and unbreakable, so it overflowed, rather than leave large spaces between words. There is also a command in tex/latex, that allows the paragraph to break, but have large spaces between words, if there is no other option.

asmeurer committed:
Split a sentence

asmeurer committed:
Wrap some text in the source

Upabjojr commented:
Will do it tomorrow morning, when you in America are asleep.

asmeurer committed:
Use texttt instead of verb for some code

This allows LaTeX to break the line in the middle.

asmeurer committed:
Either is only for two items

asmeurer committed:
Fix some wording

asmeurer committed:
Wording

asmeurer committed:
Remove first person

asmeurer committed:
Put methods of gamma in the same order they are described in the paragraph

asmeurer committed:
Put the gamma example in a figure

Not sure if figure is the correct environment for this.

asmeurer committed:
Wording

asmeurer committed:
Less colloquial language

asmeurer commented:
Should the features table contain complete sentences or not? For instance, we currently have

> Calculus: Algorithms for computing derivatives, integrals, and limits.
> Code Generation: Enables generation of compilable and executable code in a variety of different programming languages directly from expressions. Target languages include C, Fortran, Julia, JavaScript, Mathematica, Matlab and Octave, Python, and Theano.

which is inconsistent. We should either have

> Calculus: Contains algorithms for computing derivatives, integrals, and limits.
> Code Generation: Enables generation of compilable and executable code in a variety of different programming languages directly from expressions. Target languages include C, Fortran, Julia, JavaScript, Mathematica, Matlab and Octave, Python, and Theano.

or

> Calculus: Algorithms for computing derivatives, integrals, and limits.
> Code Generation: Generation of compilable and executable code in a variety of different programming languages directly from expressions. Target languages include C, Fortran, Julia, JavaScript, Mathematica, Matlab and Octave, Python, and Theano.

(note the differences at the beginning of the descriptions)

asmeurer committed:
It's not all simplification functions

asmeurer committed:
Add an example for each item in the simplification table

asmeurer committed:
Add an example for subs

asmeurer committed:
Remove finished TODO

asmeurer committed:
Remove unnecessary sentence

asmeurer committed:
Simplify some wording

asmeurer committed:
Add a comma

asmeurer committed:
Wording

Upabjojr committed:
some restructuring of sentences

Upabjojr committed:
minor

Upabjojr committed:
functional differential geometry citation

asmeurer committed:
Grammar fix

asmeurer committed:
Fix typo

asmeurer committed:
Grammar fix

asmeurer committed:
Spelling

asmeurer committed:
Wording

asmeurer committed:
Merge branch 'master' into reviews

asmeurer committed:
Merge branch 'master' into reviews

asmeurer committed:
Oxford comma

certik committed:
Merge branch 'master' into reviews

certik commented:
I merged here the latest master, which now includes the authors.

asmeurer committed:
Some small fixes to polys section

- Grammar (add "the", some commas)
- Note where we need some citations

asmeurer committed:
Clean up the polys examples a little bit

asmeurer committed:
Merge branch 'reviews' of github.com:sympy/sympy-paper into reviews

asmeurer commented:
@mattpap I pushed some small changes to the polys section here. It was nothing major (just small grammar fixes). The content looks good. However, there are a few places that need citations. Can you add citations for those (I've added comments), as well as a citation for your master's thesis. I've already imported your bib file into our bib file, so the citations you need should already be there.

Upabjojr committed:
add sympy.stats

isuruf committed:
test polys

mrocklin committed:
Pass over arch

asmeurer committed:
The SIAM style guide says to lowercase "section"

asmeurer committed:
Lowercase more instances of "section"

asmeurer committed:
Fix footnote formatting to match SIAM style guide

asmeurer commented:
@mrocklin the extra spaces around the examples are bad. This makes LaTeX think that the text after each example starts a new paragraph, like

<img width="717" alt="screen shot 2016-05-18 at 2 49 04 pm" src="https://cloud.githubusercontent.com/assets/71486/15371041/9fde52e6-1d07-11e6-95e6-baeb7908ddc5.png">

asmeurer committed:
Remove empty lines around verbatim blocks that shouldn't start new paragraphs

mrocklin commented:
My apologies.  I see that you've resolved the issue.

asmeurer committed:
Use a listing instead of a figure for the gamma example

certik committed:
Merge branch 'master' into reviews

asmeurer committed:
Wrap a long line in the polys examples

asmeurer committed:
Merge branch 'reviews' of github.com:sympy/sympy-paper into reviews

asmeurer committed:
Newline

asmeurer committed:
Describe the str printer

asmeurer committed:
Uppercase Notebook and Qt Console

asmeurer committed:
Start to clean up the solvers section

asmeurer committed:
Explain and fix the solveset domain parameter

asmeurer committed:
Remove itemize from solvers section

asmeurer committed:
Shut up chktex

asmeurer committed:
Add note about solve assuming expressions equal 0

asmeurer committed:
Include the variable in all solve examples

sumith1896 commented:
Added GSoC to acknowledgements. Can improve with better wording and the actual number of slots SymPy has received over the years.

sumith1896 committed:
Add GSoC to acknowledgement

sumith1896 committed:
Merge branch 'reviews' of github.com:sympy/sympy-paper into reviews

asmeurer committed:
Wording

asmeurer committed:
Remove the solvers section table

mattcurry committed:
Small grammatical correction.

certik commented:
The branch is getting pretty long, @asmeurer do you want to merge it into master (but keep the branch intact)? I am hoping that github is clever enough to only start showing commits + diffs from the point where you merge back into master.

asmeurer commented:
Yeah let's merge it every once in a while. The only way to do it is to merge the PR, meaning we'd have to start a new one each time, but that's fine.

mattcurry committed:
Removed unnecessary word.

mattcurry committed:
Removed comma for flow.

mattcurry committed:
Fixed some sentences in the "Features" table.

mattcurry committed:
Fixed small grammatical issue.

mattcurry committed:
Added comma.

mattcurry committed:
Fixed spelling and consistency.

mattcurry committed:
Removed redundant subsection heading.

mattcurry committed:
Removed implied word and made sentence more consistent.

mattcurry committed:
Fixed a word.

mattcurry committed:
Added new "features" table entry for quantum mechanics.

mattcurry commented:
I've read through the first half of the paper and fixed mainly grammar and flow issues. I've also added a new table entry in the "features" section for quantum mechanics.

fabianp commented:
I don't know how it is for others but I have compiling errors because of missing `pprint.pdf`

flacjacket commented on domain_specific.tex:
```diff
@@ -4,8 +4,6 @@
 mechanics along with support for manipuating physical quantities with units.
 
 
-\subsection{Classical Mechanics}
```
This should not have been removed, the below levels are sub-sub-section, this is the relevant sub-section

mattcurry replied:
My bad! I'll fix it on the next PR.

certik commented:
Is Travis passing?

Sent from my mobile phone.
On May 19, 2016 9:33 AM, "Fabian Pedregosa" notifications@github.com
wrote:

> I don't know how it is for others but I have compiling errors because of
> missing pprint.pdf
> 
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> https://github.com/sympy/sympy-paper/pull/144#issuecomment-220362497

asmeurer commented:
You need to use `make` to build the pdf.

asmeurer committed:
Merge branch 'master' into reviews

asmeurer commented:
Going to merge and re-PR as soon as Travis passes. Continue pushing to the same branch.

mattcurry committed:
Added "classical mechanics" subsection back.

flacjacket committed:
Add spaces around em dash

flacjacket committed:
Remove double period

asmeurer committed:
Fix a long line to wrap properly

asmeurer committed:
Merge branch 'reviews' of github.com:sympy/sympy-paper into reviews

flacjacket committed:
Use latex umlauts for Grobner

flacjacket committed:
Expand contraction

flacjacket committed:
Use amsmath style dots

Commit:
Expand more contractions

scopatz committed:
added my commit

scopatz committed:
scopatz intro 1

scopatz committed:
scopatz intro 2

flacjacket committed:
Use more Latex-style accents

scopatz committed:
scopatz intro 3

scopatz committed:
scopatz intro 4

scopatz committed:
scopatz intro -1

scopatz committed:
scopatz arch 0

scopatz committed:
scopatz arch 1

scopatz committed:
scopatz arch 2

scopatz committed:
scopatz arch 3

scopatz committed:
scopatz arch 4

scopatz committed:
scopatz arch 5

scopatz committed:
scopatz arch 6

scopatz committed:
Merge branch 'reviews' of github.com:sympy/sympy-paper into reviews

asmeurer committed:
Spelling

asmeurer committed:
Wording

asmeurer committed:
Merge branch 'reviews' of github.com:sympy/sympy-paper into reviews

Commit:
Python doesn't use overloading, change wording

asmeurer committed:
This is discussed below

asmeurer committed:
Move numerical simplification section to the supplement

flacjacket committed:
No space between period and footnote

scopatz committed:
scopatz arch 7

scopatz committed:
Merge branch 'reviews' of github.com:sympy/sympy-paper into reviews

asmeurer commented on architecture.tex:
```diff
@@ -194,6 +194,8 @@ \subsection{Logical Inference and Assumptions}
 \verb|(x + y).is_positive| will be \texttt{True}.
 
+%% TODO: axe this? Why mention it if we won't discuss it. Maybe move to
+%% future work. - scopatz
```
I mentioned it because people who know about it would be wondering about it. But putting it in the future work section sound like a good idea. I'll add something now.

scopatz committed:
scopatz arch 8

asmeurer committed:
The assumptions section isn't about logical inference

scopatz committed:
scopatz arch 9

scopatz committed:
scopatz arch 10

scopatz committed:
scopatz poly 1

scopatz committed:
scopatz pprint 0

scopatz committed:
scopatz pprint 1

asmeurer committed:
Merge branch 'reviews' of github.com:sympy/sympy-paper into reviews

asmeurer committed:
Move discussion of new assumptions to the future work section

scopatz committed:
scopatz solve 0

scopatz committed:
Merge branch 'reviews' of github.com:sympy/sympy-paper into reviews

scopatz committed:
scopatz solve 1

flacjacket committed:
Minor printer fixes

scopatz committed:
scopatz mat 0

scopatz committed:
Merge branch 'reviews' of github.com:sympy/sympy-paper into reviews

scopatz committed:
scopatz num 0

scopatz committed:
scopatz ds 1

scopatz committed:
scopatz fw 0

asmeurer committed:
-therefore

asmeurer committed:
Merge branch 'reviews' of github.com:sympy/sympy-paper into reviews

asmeurer commented on matrices.tex:
```diff
@@ -25,6 +25,8 @@
 \end{verbatim}
 
+%% TODO a Python list is not a dense representation...
+%% Is this supposed to say flat? - scopatz
```
Yes it is. If you have a 1000 x 1000 matrix, the list representation requires representing all 1000000 elements (even if most of them are zeros).

scopatz replied:
I understand that it is a dense matrix, but the list isn't a dense data structure. A list of lists of symbols would still be a dense matrix but the representation would not be any less dense as a data structure (in Python). I think this could be reworded because it implies that the flatness of the list makes the matrix dense, which is not true.

asmeurer replied:
How is it not a dense data structure?

asmeurer replied:
It isn't really flat either. Does this help? https://github.com/sympy/sympy-paper/pull/144/commits/332203b9ce19aeecf0b703be1f5284c85113b0f9

scopatz replied:
Well a list is collection of `PyObject*`, [namely](http://www.laurentluce.com/posts/python-list-implementation/), 

``` c
typedef struct {
    PyObject_VAR_HEAD
    PyObject **ob_item;
    Py_ssize_t allocated;
} PyListObject;
```

The individual expression elements are all in their own locations on the heap. This is about as non-dense as it gets in terms of memory allocation. 

> It isn't really flat either. Does this help? 332203b

It isn't flat because it is a list of lists? That is fine too. I guess what I am trying to say is that the denseness and the data structure should be independent statements.  Something like, "The Matrix class represents dense matrices. Internally, these matrices store their elements as a list of lists."

asmeurer replied:
You and I are working with different definitions of sparse and dense. The one used here is the same as Jason Grout defines it [here](http://ask.sagemath.org/question/9464/what-is-the-difference-between-dense-and-sparse-matrices/), namely, dense means all values are stored, and sparse means only nonzero values are stored.  Perhaps we should use the specific terminology [LIL](https://en.wikipedia.org/wiki/Sparse_matrix#List_of_lists_.28LIL.29). 

The same usage is used in the polynomials section, by the way,

>  SymPy implements dense and sparse polynomial representations. Both are used in the univariate and multivariate cases. The dense representation is the default for univariate polynomials. For multivariate polynomials, the choice of representation is based on the application. The most common case for the sparse representation is algorithms for computing Gröbner bases (Buchberger, F4, and F5). This is because different monomial orderings can be expressed easily in this representation. However, algorithms for computing multivariate GCDs or factorizations, at least those currently implemented in SymPy, are better expressed when the representation is dense. The dense multivariate representation is specifically a recursively dense representation, where polynomials in K[x0, x1, . . . , xn] are viewed as a polynomials in K[x0][x1] . . . [xn]. Note that despite this, the coefficient domain K, can be a multivariate polynomial domain as well. The dense recursive representation in Python gets inefficient when the number of variables gets high.

scopatz replied:
I don't think we are using different definitions when it comes to the matrices themselves. We both agree that a dense matrix stores all elements.  That said, I agree that we should be as specific as possible.  The LIL that you cited didn't mention a particular language. I'll add that Python is the weird case here where lists are not densely or contiguously stored.  

The polynomial usage is fine because it doesn't tie the densely causally to a Python data structure, which can be said to be non-dense. 

I think the matrix case could be a place where a reviewer would find this objectionable. It is better to be non-ambiguous and prevent such objections.

asmeurer replied:
It sounds like you're conflating dense and contiguous. Python lists may not be contiguous, but they are dense. `[0]*1000` takes (amortized) 1000 as much space to store as `[0]` (same with `[None]*1000`). Javascript arrays (from my understanding of them) are an example of a non-dense array. You can have a Javascript array where the first 1000 items are `undefined` and the 1001st item is `1` and it takes up as much space as `[1]`. The closest thing to a non-dense array in Python is a dictionary indexed by non-negative integers (except you'd need a wrapper because all the methods like len and iteration are wrong).  

If I Google "dense data structure" I only get results that are related to matrices or polynomials. If I Google "dense array" I get references to the non-denseness of Javascript arrays.  Based on that I'd say this is a standard usage of the term. 

I'm all for adding clarity. Should we define "dense" and "sparse"?

scopatz replied:
I can find a number of places where people refer to dense memory allocation when they refer to contiguous chunks:
[here](https://books.google.com/books?id=DqX-aWwIEJsC&pg=PA411&lpg=PA411&dq=dense+memory+allocation&source=bl&ots=a4r5UlBW9u&sig=46mXpruUj7pY4F5_HxoaR9dzvAI&hl=en&sa=X&sqi=2&ved=0ahUKEwj2q8O7sufMAhXDSiYKHf0cBx8Q6AEIKzAD#v=onepage&q=dense%20memory%20allocation&f=false) and [here](https://wiki.dlang.org/Dense_multidimensional_arrays#Dense_dynamic_arrays) seem relevant to the matrix discussion.  I don't think I am the only one who may conflates these as I have heard it used as a term of art, though personally, mostly in database discussions.

Also, I do think that the Python array is a dense/contiguous data structure. Also, if I am not mistaken, slots, strings, bytes, and structs are also all dense/contiguous in their storage. 

Defining the terms and then using them clearly is all that I desired out of my original comment.

scopatz replied:
Also, here is another document that uses the term dense w.r.t. to contiguous memory https://www.kernel.org/doc/ols/2007/ols2007v1-pages-141-152.pdf

asmeurer replied:
I added a footnote to the polys section explaining what they mean (polys because the polys section comes before the matrix section). Hopefully it is sufficient to also clarify the term for the subsequent matrices section, even though the strict definition is slightly different (a matrix has entries and shape; a polynomial has coefficients and degree).

scopatz replied:
I like the footnote, but I would have to read the whole paper through again to see if that section is still in my L1 or L2 by the time I get to this section.  Since the "makes it" wording is still in there, I think that a comment tying this back would make this more explicit, "dense (in the same sense as a "dense polynomial")"., or make that a footnote.

scopatz replied:
Also, this is something that I can do on my next read through

asmeurer committed:
Spelling

asmeurer committed:
Spelling

asmeurer committed:
MATLAB is all upppercase (according to SIAM style guide)

asmeurer committed:
bra-ket should use en dash

asmeurer committed:
built in (spelling)

asmeurer committed:
Spelling

asmeurer committed:
Spelling

asmeurer committed:
Fix some LaTeX formatting in other_projects_that_use_sympy.tex

asmeurer committed:
Lists are dense

See
https://github.com/sympy/sympy-paper/pull/144/commits/57a12cbf3aef00df64d66cad0126f6178484e2fe#r63944359

asmeurer committed:
List of lists

asmeurer committed:
Make the description column of features table larger

This makes it easier to read, and take up less vertical space.

asmeurer committed:
Move polys, solvers, and matrices examples to the supplement

This is because of space constraints. The paper is already too long.

asmeurer committed:
Move simplification examples to the supplement

mattcurry committed:
Fixed a sentence.

mattcurry committed:
Enhanced a qubit example.

asmeurer committed:
Define dense and sparse in a footnote

asmeurer committed:
Merge branch 'reviews' of github.com:sympy/sympy-paper into reviews

fredrik-johansson committed:
condense the numerics section

fredrik-johansson committed:
numerics: more editing; move comparison with mma to comaprison_with_mma

fredrik-johansson commented:
I've edited my own contributions a bit; please review the changes I made.

I've skimmed most of the rest of the paper, and it looks pretty good. Not sure if I will have time to read everything very carefully.

asmeurer committed:
Fix doctest failure

asmeurer committed:
Add reference to the supplement in the limits section

asmeurer commented:
Travis finally caught, so I merged this. Continued at https://github.com/sympy/sympy-paper/pull/158.
